### PR TITLE
get Pagination data before Models

### PR DESCRIPTION
**Pull request by lav45**
``` php
    public function actionIndex()
    {
        $_GET['page'] = 3;

        $dataProvider = new ArrayDataProvider([
            'allModels' => [
                ['id' => 1],
                ['id' => 2],
                ['id' => 3],
                ['id' => 4],
                ['id' => 5],
                ['id' => 6],
                ['id' => 7],
            ],
            'pagination' => [
                'pageSize' => 2,
            ],
        ]);

        var_dump($dataProvider->getPagination()->getPage());
        // return => int 0

        var_dump($dataProvider->getModels());
        // return => array (size=2)
        //  0 => 
        //    array (size=1)
        //      'id' => int 1
        //  1 => 
        //    array (size=1)
        //      'id' => int 2

        return 'OK';
    }
```

``` php
    public function actionIndex()
    {
        $_GET['page'] = 3;

        $dataProvider = new ArrayDataProvider([
            'allModels' => [
                ['id' => 1],
                ['id' => 2],
                ['id' => 3],
                ['id' => 4],
                ['id' => 5],
                ['id' => 6],
                ['id' => 7],
            ],
            'pagination' => [
                'pageSize' => 2,
            ],
        ]);

        var_dump($dataProvider->getModels());
        // return => array (size=2)
        //  4 => 
        //    array (size=1)
        //      'id' => int 5
        //  5 => 
        //    array (size=1)
        //      'id' => int 6

        var_dump($dataProvider->getPagination()->getPage());
        // return => int 2

        return 'OK';
    }
```

### CODE REVIEW

File: tests/framework/data/ActiveDataProviderTest.php [hunk 1]
```diff
@@ -168,7 +170,6 @@ public function testPaginationBeforeModels()
             'query' => $query->from('order')->orderBy('id'),
         ]);
         $pagination = $provider->getPagination();
-        $this->assertEquals(0, $pagination->getPageCount());
```
Comment (LAV45) [inline tests/framework/data/ActiveDataProviderTest.php]: This line also points to this problem

**Comment by LAV45** [comment]
And here is the source of the problem https://github.com/LAV45/yii2/blob/master/framework/data/Pagination.php#L191

**Comment by xjflyttp** [comment]
I think you call before $dataProvider->prepare();

**Comment by qiangxue** [comment]
The change is not correct. It will make the pagination configuration to be sensitive about whether pagination is configured first or not.

As described in the API comment, you should call `prepare()` or `getModels()` before calling `getPagination()`.

**Comment by LAV45** [comment]
@qiangxue,  If the api class has methods, procedure call, which may affect the result obtained, it may be worthwhile to protect such methods from external access?

**Pull request closed without merging**